compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -225,6 +231,22 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         }
     }
 
+    fn ptr_to_metadata(
+        &self,
+        src: &ImmTy<'tcx, M::Provenance>,
+        cast_to: TyAndLayout<'tcx>,
+    ) -> InterpResult<'tcx, ImmTy<'tcx, M::Provenance>> {
+        assert!(src.layout.ty.is_unsafe_ptr());
+        return Ok(match **src {
+            Immediate::Scalar(_) => {
+                assert!(cast_to.is_zst());
+                ImmTy::uninit(cast_to)
+            }
+            Immediate::ScalarPair(_, meta) => ImmTy::from_scalar(meta, cast_to),
+            Immediate::Uninit => throw_ub!(InvalidUninitBytes(None)),
```

Comment:
Unsure whether we want to raise UB here. We don't raise UB for offsetting an immediate to get a sub-immeidate

Reply:
either way, needs const-eval and miri tests

Reply:
An uninit thin pointer violates the thin pointer validity invariant. I don't think we have any case here that we want to allow -- all our pointers are `noundef` on both the data and metadata component.